전남대 BE_이장안 2주차 과제 (0단계)

lja3723 opened the pull request:
## Week2 Pull Request List

- 다음을 참고바랍니다.
- https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pulls?q=is%3Apr+author%3Alja3723

## Change Log

- 기본 변경사항
  - 이전 과제의 피드백을 반영하였습니다. ([로그 확인하기](
https://github.com/lja3723/spring-gift-wishlist/compare/b1fc03be448e231424554c6826689ab1e916ecc7...946c57aa6f76fde2ab937cf3069fa0928b173569))
- 피드백 반영 후 변경사항
  - 피드백 반영하여 Dto를 옮기는 중, 앞으로 과제 수행을 생각하여 패키지 구조를 변경하였습니다. ([참고자료](https://gksdudrb922.tistory.com/271)) ([로그 확인하기](https://github.com/lja3723/spring-gift-wishlist/compare/b19661ede51c8d7aba0f8b5ec127607fad89c6a9...859a79ee724cfe3af116f25a15d1f148137e2859), [로그 확인하기 2](https://github.com/lja3723/spring-gift-wishlist/commit/859a79ee724cfe3af116f25a15d1f148137e2859))
    - 계층형 패키지 구조 -> 도메인형 패키지 구조로 변경
  - 일관성 있는 응답을 위해 API 명세서를 수정하였습니다. ([로그 확인하기](https://github.com/lja3723/spring-gift-wishlist/compare/7f4bfed86c932822008c374708f13edae1659d90...5ca84b11b936e4ec082e0fba3070abbc3664239c))
    - 모든 응답에는 `timestamp`, `status` 키가 기본적으로 포함됨
    - API에 따라 추가적인 키가 포함될 수 있음

## TODO

- Logger의 사용은 추후 반영하겠습니다.
- 테스트 코드 추가하도록 하겠습니다.

lja3723 committed:
chore: migrate sources from https://github.com/lja3723/spring-gift-product

이전 과제 https://github.com/lja3723/spring-gift-product 로부터 코드를 가져옴

lja3723 committed:
refactor: apply previous code review feedback (1)

이전 과제의 코드리뷰 피드백을 반영함
- 레포지토리의 공통 모듈(RowMapper) 메서드화
- 파라미터 변수 이름 구체화
- Optional 사용 관련 최적화
- 서비스 메서드 이름 구체화
- HTML 마지막 행 공백 삽입(컨벤션에 의함)

lja3723 committed:
refactor: apply previous code review feedback (2)

이전 과제의 코드리뷰 피드백을 반영함
- API 명세서 변경
  - 이제는 상품 추가 시 추가된 정보가 응답으로 반환됨
- DTO 이름 변경

lja3723 committed:
refactor: apply previous code review feedback (3)

이전 과제의 코드리뷰 피드백을 반영함
- Controller <-> Service <-> Repository 레이어 간 책임 분리
  - 이제는 Controller는 Dto에만 의존함
  - Service에서 Dto <-> Entity 간 변환 담당

lja3723 committed:
refactor: apply previous code review feedback (4)

이전 과제의 코드리뷰 피드백을 반영함
- 상품 수정 API의 메서드를 PUT으로 변경
  - API 명세서에 반영
  - PUT으로 변경하기 위한 부가적인 작업 수행
    - th:method 사용, application.properties 속성 추가 등

gunom commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,25 @@
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(ProductNotFoundException.class)
+    public ResponseEntity<Void> handleProductNotFoundException(ProductNotFoundException e) {
```
void로 리턴하지 말고 이것처럼 ErrorReseponse를 직접 정의해서 내리면 더 체계적으로 에러를 핸들링할 수 있을 것 같아요~

```java
public static class ErrorResponse {
        private String message;
        private int status;
...
}
```

lja3723 replied:
조언 감사합니다! 반영했습니다.
추가적으로 PR 보낸 사항 확인해 보시고 문제 없으면 merge 부탁드립니다~ 왜냐면 merge해 주셔야 step1 진행이 가능합니다!
변경 기록 아래 링크 참고바랍니다.
https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/53#issue-2383063360

gunom replied:
안녕하세요 장안님 리뷰 반영 확인했습니다 :) 언제든 하시다가 막히시거나 질문있으시면 PR이나 구건모 멘토로 멘토링 신청해주세여 :)

lja3723 replied:
감사합니다!

gunom commented on src/main/java/gift/repository/ProductRepository.java:
```diff
@@ -0,0 +1,76 @@
+        try {
+            product =  jdbcTemplate.queryForObject(sql, getRowMapper(), id);
+        } catch (EmptyResultDataAccessException e) {}
+        return Optional.ofNullable(product);
```
catch에서 에러가 발생하면 아무것도 안하고 있는데 이 부분에서 바로 빈 값을 리턴해도 되지 않을까요?
```java
        catch (EmptyResultDataAccessException e) {
            return Optional.empty();
        }
```
이러면 product를 try 밖에서 따로 지정안하고 try 내부에서 리턴할 수 있을 것 같네요

lja3723 replied:
반영하겠습니다~

gunom reviewed (commented):
0단계 준비 잘해주셨습니다! 몇 가지 리뷰 남겨놨으니 참고해주세요~

또 하나 첨언드리자면, Product는 entity이므로 계층 상 model에 존재하는 것은 아주 좋은 판단입니다! 그러나, ProductRequestDto와 ProductResponseDto는 Contorller Layer와 Service Layer에서 사용되므로 model 패키지가 아닌 domain혹은 service나 controller 패키지로 이동되면 더 좋을 것 같습니다 :)

리뷰 중 질문있으시면 편하게 남겨주세요

lja3723 committed:
refactor: apply code review feedback (5)

코드 리뷰 피드백을 반영함
- try-catch flow 최적화

lja3723 committed:
chore: change package structure

계층형 패키지 구조에서 도메인형 패키지 구조로 변경

lja3723 committed:
refactor: separate non-RESTful exception handler from GlobalExceptionHandler

non-RESTful 한 예외 핸들러를 글로벌 예외 핸들러로부터 분리해 ServerRenderController만을 위한 예외 핸들러로 종속시켜, 글로벌 예외 핸들러를 RestControllerAdvice로 변경

lja3723 committed:
refactor: modify API response

- API response를 일관성 있게 변경
- 일관성 있는 응답을 위해 응답에 관한 여러 유틸리티 클래스 추가(SuccessResponse, ErrorResponse 등)
